src/servers/FtpServer.cc

```diff
@@ -1700,18 +1701,21 @@ Ftp::Server::checkDataConnPost() const
 void
 Ftp::Server::connectedForData(const CommConnectCbParams &params)
 {
-    connector = NULL;
+    dataConnWait.finish();
 
     if (params.flag != Comm::OK) {
-        /* it might have been a timeout with a partially open link */
-        if (params.conn != NULL)
-            params.conn->close();
         setReply(425, "Cannot open data connection.");
         Http::StreamPointer context = pipeline.front();
         Must(context->http);
         Must(context->http->storeEntry() != NULL);
     } else {
-        Must(dataConn == params.conn);
+        // Finalize the details and start owning the supplied connection.
+        assert(params.conn);
```

Comment:
I would remove the 'Must' below and add the corresponding assert for symmetry. Also, that would eliminate the duplicated 'params.conn is not nil' check Comm::IsConnOpen().
```suggestion
        assert(params.conn);
        assert(params.conn.isOpen());
```

Reply:
I think these asserts should be left as is because they check the sanity/safety of the dataConn assignment, and params.conn openness has no bearing on that.

You are right that the Must() below should be replaced with an assert, but that Must() is about the safety of the fd_note() call below it, and we are not changing that call, so I would keep the Must() as is as well, for now. Another reason to keep that Must() is to avoid an argument that we are replacing a Must() with an identical assert(). Until we add the [missing](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api) Assert() API or equivalent, it is a somewhat difficult argument to defend against.

Reply:
Yes, there is. During my next pass, I was (and still am) going to evaluate whether to add a wrapper for all those similar cases.

Reply:
I did not add a wrapper because we would probably need several -- all these use cases around the code are similar but not identical. It may still be a good idea to add it, but I did not want to add one more new API to the PR.